src/app/clusters/color-control-server/color-control-server.cpp

```diff
@@ -1528,24 +1629,39 @@ static bool computeNewHueValue(ColorHueTransitionState * p)
     }
 
     // are we going up or down?
-    if (p->finalHue == p->currentHue)
+    if ((p->isEnhancedHue && p->finalEnhancedHue == p->currentEnhancedHue) || (!p->isEnhancedHue && p->finalHue == p->currentHue))
     {
         // do nothing
     }
     else if (p->up)
     {
-        newHue32 = (uint32_t) subtractHue(p->finalHue, p->initialHue);
-        newHue32 *= ((uint32_t)(p->stepsRemaining));
-        newHue32 /= ((uint32_t)(p->stepsTotal));
-        p->currentHue = subtractHue((uint8_t) p->finalHue, (uint8_t) newHue32);
+        newHue32 = static_cast<uint32_t>(p->isEnhancedHue ? subtractEnhancedHue(p->finalEnhancedHue, p->initialEnhancedHue)
+                                                          : subtractHue(p->finalHue, p->initialHue));
+        newHue32 *= static_cast<uint32_t>((p->stepsRemaining / p->stepsTotal));
```

Comment:
This rearrangement of the arithmetic is not right.

Say we are moving hue from initial 100 to final 200.  The `subtractHue` result is 100.  Let's say we're doing it over 50 steps and we have taken 1 of those steps.  So `stepsTotal` is 50 and `stepsRemaining` is 49.  The old calculation came out with `newHue32 = (100 * 49) / 50 = 4900 / 50 = 98`.  Then we would subtract that from the final value and end up with a current hue value of 102, as expected.  And if we had taken 45 steps and had 5 stepsRemaining, we would get a current hue of 190.

The new calculation does `newHue32 = 100 * (49/50) = 100 * 0 = 0` (integer division, remember!), and comes out with a current hue of 200.  In fact, anything other than stepsRemaining == stepsTotal gives 200.

I wish we had test coverage for this, given all the changes we are making...  :(  I would hope it would have caught this problem.  Can you please make sure (at least as a followup) that tests are created?

Reply:
I see. Thank you for catching this Boris. I splitted them back how they were Multiplication first. then divide that result.

I will take actions to start a test coverage for this cluster starting next week